lisa/wlgen/workload.py

```diff
@@ -80,15 +81,29 @@ def __init__(self, target, name, res_dir=None):
         self.name = name
         self.command = None
         self.output = ""
-        self.run_dir = self.target.working_directory
 
+        self.target.execute(
+            "mkdir -p {}/lisa/wlgen".format(target.working_directory))
+        self.run_dir = target.execute(
+            "mktemp -d -p {}/lisa/wlgen {}_{}_XXXXXX".format(
+                self.target.working_directory, self.name,
+                datetime.now().strftime('%Y%m%d_%H%M%S'))).strip()
+        
+        logger = self.get_logger()
+        logger.info("Creating target's run_dir: %s", self.run_dir)
+        
         res_dir = res_dir if res_dir else target.get_res_dir(
             name='{}-{}'.format(self.__class__.__qualname__, name)
         )
         self.res_dir = res_dir
 
         self.target.install_tools(self.required_tools)
 
+    def __del__(self):
```

Comment:
Using `__del__` is a last resort measure as there are many caveats and surprising things that can happen there (I won't go in the details here but calling some APIs like devlib target is definitely far from what can be expected to work flawlessly).

If there are some resources to cleanup, that needs to go in a context manager instead.

Reply:
It seems quite convoluted to use a context manager in this context.

We need to create the `res_dir` when a `workload` is created and before `run()` is actually used, since the client perhaps wants to do something with that folder, e.g. push custom assets. 

Moreover, for the same reason above, we don't really know when it's safe to wipe the output folder. Thus, I would I just turn the `dtor` into a normal method and give the caller the chance to use it if and when it needs it.

Reply:
You can provide a close() method and implement `__exit__` and `__enter__` to use it. Calling `close()` explicitly might indeed be needed in some non-classic cases, but requires lot of special care to work properly. Context managers will provide something robust all the time with no boilerplate to add for client code, and therefore must be used 1) whenever it's possible in existing code, and 2) new client code must be designed so it uses them.

Reply:
I looked into the possibility to turn this into a context manager but its usage looks not so straightforward in our current code. Indeed, it cannot be used in `RTATestBundle::run_rtapp()`
   https://github.com/ARM-software/lisa/blob/master/lisa/tests/base.py#L1136
the same way we already do for the others, i.e. in:
```python
with dmesg_coll, ftrace_coll, target.freeze_userspace():
            wload.run(cgroup=cgroup, as_root=as_root)
```
since when we return from `run_rtapp` we still want to give a chance to the caller to pull data from the target.

We should probably use such a context manager from the `TestBundle` base class.
But that's a much more invasive change and I rather prefer to leave it for a future change.

So, for the time being, since there are not actual usages, I would prefer to avoid adding useless `__exit__` and `__enter__` calls.

Reply:
I've introduced a `wipe_res_dir()` in the last push which I still find it useful at least when I run tests from my notebooks.

However, if you are not convinced without a full context manager,  I'm also happy to drop it.

Reply:
In the current state, `run_rtapp()` could use the context manager very easily by increasing the indentation of this code by one level, so that it sits inside the `with` statement:
```
        ftrace_coll.get_trace(trace_path)
        dmesg_coll.get_trace(dmesg_path)
```

If somebody needs to have access to the folder after the execution, we should add a `wipe_run_dir=True` parameter to `run_rtapp()`, so that people can disable it if they need and then bracket it properly with try/except. However, that's not the usual case and no test currently need access to that AFAIK.